cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceRuntimeService.java

```diff
@@ -504,13 +506,20 @@ private void configureStreamInput(Job job, StreamBatchReadable stream) throws IO
     StreamInputFormat.setStreamPath(job, streamPath.toURI());
     StreamInputFormat.setTimeRange(job, stream.getStartTime(), stream.getEndTime());
 
-    String decoderType = stream.getDecoderType();
-    if (decoderType == null) {
-      // If the user don't specify the decoder, detect the type from Mapper/Reducer
-      setStreamEventDecoder(job);
+    FormatSpecification formatSpecification = stream.getBodyFormatSpec();
+    if (formatSpecification != null) {
```

Comment:
Can you open a JIRA for the case when the format spec is available but the mapper/reducer is using a different type, we should be able to do type projection into the target type.